modules/tensor_mechanics/src/materials/ComputeLinearElasticPFFractureStress.C

```diff
@@ -8,112 +8,247 @@
-                                          getVar("c", 0)->name())),
-    _d2Fdc2(declarePropertyDerivative<Real>(
-        getParam<MaterialPropertyName>("F_name"), getVar("c", 0)->name(), getVar("c", 0)->name())),
-    _d2Fdcdstrain(declareProperty<RankTwoTensor>("d2Fdcdstrain")),
-    _dstress_dc(declarePropertyDerivative<RankTwoTensor>("stress", getVar("c", 0)->name())),
-    _hist(declareProperty<Real>("hist")),
-    _hist_old(getMaterialPropertyOld<Real>("hist")),
-    _kappa(declareProperty<Real>(getParam<MaterialPropertyName>("kappa_name"))),
-    _L(declareProperty<Real>(getParam<MaterialPropertyName>("mobility_name")))
+  : ComputePFFractureStressBase(parameters),
+    GuaranteeConsumer(this),
+    _decomposition_type(getParam<MooseEnum>("decomposition_type"))
 {
 }
 
 void
-ComputeLinearElasticPFFractureStress::computeQpStress()
+ComputeLinearElasticPFFractureStress::initialSetup()
+{
+  if ((_decomposition_type == "strain_vol_dev" || _decomposition_type == "strain_spectral") &&
+      !hasGuaranteedMaterialProperty(_elasticity_tensor_name, Guarantee::ISOTROPIC))
+    mooseError("Decomposition approach of strain_vol_dev and strain_spectral can only be used with "
+               "isotropic elasticity tensor materials, use stress_spectral for anistropic "
+               "elasticity tensor materials");
+}
+
+void
+ComputeLinearElasticPFFractureStress::computeStrainSpectral(Real & F_pos, Real & F_neg)
 {
-  const Real c = _c[_qp];
+  // Isotropic elasticity is assumed and should be enforced
+  const Real lambda = _elasticity_tensor[_qp](0, 0, 1, 1);
+  const Real mu = _elasticity_tensor[_qp](0, 1, 0, 1);
+
+  // Compute eigenvectors and eigenvalues of mechanical strain and projection tensor
+  RankTwoTensor eigvec;
+  std::vector<Real> eigval(LIBMESH_DIM);
+  RankFourTensor Ppos =
+      _mechanical_strain[_qp].positiveProjectionEigenDecomposition(eigval, eigvec);
+  RankFourTensor I4sym(RankFourTensor::initIdentitySymmetricFour);
+  RankFourTensor Pneg = I4sym - Ppos;
 
-  // Zero out values when c > 1
-  Real cfactor = 1.0;
-  if (c > 1.0)
-    cfactor = 0.0;
+  // Calculate tensors of outerproduct of eigen vectors
+  std::vector<RankTwoTensor> etens(LIBMESH_DIM);
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+        etens[i](j, k) = eigvec(j, i) * eigvec(k, i);
+
+  // Separate out positive and negative eigen values
+  std::vector<Real> epos(LIBMESH_DIM), eneg(LIBMESH_DIM);
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    epos[i] = (std::abs(eigval[i]) + eigval[i]) / 2.0;
+    eneg[i] = (std::abs(eigval[i]) - eigval[i]) / 2.0;
+  }
+
+  // Seprate positive and negative sums of all eigenvalues
+  Real etr = 0.0;
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    etr += eigval[i];
+
+  const Real etrpos = (std::abs(etr) + etr) / 2.0;
```

Comment:
I could be wrong, but I think std::max(etr,0.0) is faster and more accurate than this line here.